src/clarity-icons/clarity-icons-element.ts

```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
+    // before this connectedCallback method called only once when the custom element is inserted into the DOM.
+    // So we could check whether the attribute values really changed or not.
+    // If not, we don't need to execute the same codes again.
     const host = this as HTMLElement;
 
     if (host.hasAttribute("shape")) {
-        generateIcon(host, host.getAttribute("shape"));
+        const shapeAttrValue = host.getAttribute("shape").split(/\s/)[0];
+
+        if (shapeAttrValue === this.currentShapeAttrVal) {
+            return;
+        }
+
+        this.currentShapeAttrVal = shapeAttrValue;
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+
+    if (host.hasAttribute("title")) {
+        const titleAttrValue = host.getAttribute("title");
+
+        if (this.currentTitleAttrVal !== titleAttrValue) {
+            this.currentTitleAttrVal = titleAttrValue;
+        }
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
     }
 
     if (host.hasAttribute("size")) {
-        setIconSize(host, host.getAttribute("size"));
+        const sizeAttrValue = host.getAttribute("title");
+
+        if (this.currentSizeAttrVal !== sizeAttrValue) {
+            this.currentSizeAttrVal = sizeAttrValue;
+            setIconSize(host, host.getAttribute("size"));
+        }
     }
+
+    this.injectTemplate(host);
 };
 
 ClarityIconElement.prototype.attributeChangedCallback = function(attributeName: string, oldValue: string,
                                                                  newValue: string) {
     const host = this as HTMLElement;
 
+    if (!this.initialized) {
+        this.clrIconUniqId = "_clr_icon_" + clrIconId;
+        clrIconId++;
+        this.currentShapeAttrVal = "";
+        this.currentShapeTemplate = "";
+        this.currentTitleAttrVal = "";
+        this.currentSizeAttrVal = "";
+        this.initialized = true;
+    }
+
     if (attributeName === "shape") {
-        generateIcon(host, newValue);
+        this.currentShapeAttrVal = newValue.split(/\s/)[0];
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+    if (attributeName === "title") {
+        this.currentTitleAttrVal = newValue;
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
```

Comment:
That return I understand slightly more, but even then I don't see the point. And why is this only for the title and not the size? They should have exactly the same problems with no shape.

Just fallback "gracefully" to the error icon with the correct title and size, it's the most friendly behavior and we can have it for free.

Reply:
I have moved the size testing part to the area above both the shape attribute check. So we get hold of the size attribute before shape and title attributes.

As to falling back to the error icon, in cases when title attribute value happens to change right before shape attribute changes, why do we need to inject the error icon?

Reply:
The order doesn't really matter, what I'm saying right now is that if I put a correct title but a wrong shape, I get an error icon _without_ my title. You should at least display the error icon with the correct title.
You noted this was a good idea and said you would do it, but I still don't see it anywhere.

Reply:
Wait 🤔  I've actually addressed that issue. I have even added a test for the case:
```
it("should inject error shape with title if icon is not found", ...);
```